src/shogun/machine/TypedMachine.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Shubham Shukla
+ */
+
+#ifndef _TYPEDMACHINE_H__
+#define _TYPEDMACHINE_H__
+
+#include <shogun/lib/config.h>
+
+#include <shogun/base/progress.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/common.h>
+#include <shogun/machine/LinearMachine.h>
+
+namespace shogun
+{
+	class CFeatures;
+	class CLabels;
+	
+	template <class P, class T>
+	class CTypedMachine : public T
+	{
+	public:
+		/** Default constructor */
+		CTypedMachine() : T()
+		{
+		}
+
+		virtual ~CTypedMachine()
+		{
+		}
+
+	protected:
+		virtual bool train_machine(CFeatures* data = NULL)
+		{
+				if (!data)
+				{	
+					REQUIRE(this->features, "No features provided.\n")	
+					REQUIRE(this->features->get_feature_class() == C_DENSE,	
+					"Feature-class (%d) must be of type C_DENSE (%d)\n", this->features->get_feature_class(), C_DENSE)	
+					
+					data = this->features;	
+				}	
+		else	
+			REQUIRE(data->get_feature_class() == C_DENSE,	
+				"Feature-class must be of type C_DENSE (%d)\n", data->get_feature_class(), C_DENSE)	
+		
+			REQUIRE(data->get_num_vectors() == this->m_labels->get_num_labels(), "Number of training vectors (%d) does not match number of labels (%d)\n"	
```

Comment:
if this class is just for dense, then such checks should be one level up in the hierarchy since we dont want to copy paste them around

Reply:
this will be general but i have only written the dense case

Reply:
not so sure about that.
Add a case for string and you will see the problem (see above)

Reply:
yes i see it :)